src/stm32f1/pwm.cpp

```diff
@@ -0,0 +1,230 @@
+      bit_modify(p_reg->cc_enable_register)
+        .insert<ccer_polarity_enable_mask_3>(0b01U);
+      break;
+    case 4:
+      static constexpr auto ccer_polarity_enable_mask_4 =
+        bit_mask::from(12, 13);
+      bit_modify(p_reg->cc_enable_register)
+        .insert<ccer_polarity_enable_mask_4>(0b01U);
+      break;
+    default:
+      break;
+  }
+
+  bit_modify(p_reg->break_and_deadtime_register)
+    .clear(ossr)
+    .set(main_output_enable);  // complementary channel stuff
+
+  return;
+}
+void setup(pwm_pins p_pin)
+{
+
+  static constexpr auto clock_division = bit_mask::from<8, 9>();
+  static constexpr auto edge_aligned_mode = bit_mask::from<5, 6>();  // set to
+                                                                     // 00
+  static constexpr auto direction = bit_mask::from<4>();
+
+  static constexpr auto output_compare_odd = bit_mask::from<4, 6>();
+  static constexpr auto output_compare_even = bit_mask::from<12, 14>();
+  static constexpr auto counter_enable = bit_mask::from<0>();  // set this to 1
+  static constexpr auto odd_channel_preload_enable = bit_mask::from<3>();
+  static constexpr auto even_channel_preload_enable = bit_mask::from<11>();
+
+  peripheral peripheral_id = get_peripheral_id(p_pin);
+  pwm_reg_t* reg = get_pwm_reg(peripheral_id);
+
+  bit_modify(reg->control_register).insert<clock_division>(0b00U);  //
+  bit_modify(reg->control_register).insert<edge_aligned_mode>(0b00U);
+  bit_modify(reg->control_register).clear(direction);
+
+  if (p_pin == pwm_pins::pa8) {  // channel 1
+    // Preload enable must be done for corresponding OCxPE
+    bit_modify(reg->capture_compare_mode_register)
+      .insert<output_compare_odd>(0b110U);
+    bit_modify(reg->capture_compare_mode_register)
+      .set(odd_channel_preload_enable);
+    setup_channel(reg, 1);
+  } else if (p_pin == pwm_pins::pa9) {  // channel 2
+    bit_modify(reg->capture_compare_mode_register)
+      .insert<output_compare_even>(0b110U);
+    bit_modify(reg->capture_compare_mode_register)
+      .set(even_channel_preload_enable);
+    setup_channel(reg, 2);
+
+  } else if (p_pin == pwm_pins::pa10) {  // channel 3
+    bit_modify(reg->capture_compare_mode_register_2)
+      .insert<output_compare_odd>(0b110U);
+    bit_modify(reg->capture_compare_mode_register_2)
+      .set(odd_channel_preload_enable);
+    setup_channel(reg, 3);
+
+  } else if (p_pin == pwm_pins::pa11) {  // channel 4
+    bit_modify(reg->capture_compare_mode_register_2)
+      .insert<output_compare_even>(0b110U);
+    bit_modify(reg->capture_compare_mode_register_2)
+      .set(even_channel_preload_enable);
+    setup_channel(reg, 4);
+  }
+
+  bit_modify(reg->control_register).set(counter_enable);
+
+  reg->prescale_register = 0x0U;
+
+  // Set counter value to 0 because it upcounts.
+  reg->counter_register = 0x0U;
+
+  //
+  reg->auto_reload_register = 0xFFFF;
+}
+
+}  // namespace
+
+pwm::pwm(pwm_pins p_pin)
```

Comment:
If this class only works for timer1, then we should rename this class to `timer1_pwm` so its clear.